Make `ismutable` a type-parameter of `FunctionMap`

dkarrasch opened the pull request:
This takes a small step towards what I think is wide-spread in the SciML/DiffEq world: making mutability a type-parameter. This allows to solve #193 by multiple dispatch, rather than querying a specific field of all factors/summands. The downside is that previous constructor calls are potentially not fully inferred, see the changed tests. If we advise users, however, to use the `FunctionMap` constructor explicitly and provide the mutability information, that is most likely statically known anyway, directly, then that shouldn't do much harm. In any case, the current design is such that the "old" behavior still works, it's just deprecated. What I find nice is that for out-of-place `FunctionMap`s, `A*B*C*D*x` first folds from the left, so that we first create a `CompositeMap` until we hit `x`, from where it folds from the right, without going through 3-arg `mul!`.

dkarrasch commented:
@JeffFessler I suspect you are using `FunctionMap`s heavily. Can you please try your packages with this PR branch and report back any new issues? I believe nothing should break, only some deprecation warnings should appear, but you never know.

JeffFessler commented on src/LinearMaps.jl:
```diff
@@ -18,8 +17,9 @@ using Base: require_one_based_indexing
 
-const MapOrVecOrMat{T} = Union{LinearMap{T}, AbstractVecOrMat{T}}
-const MapOrMatrix{T} = Union{LinearMap{T}, AbstractMatrix{T}}
+const AbstractVecOrMatOrQ{T} = Union{AbstractVecOrMat{T}, AbstractQ{T}}
```
I had never head of `AbstractQ` before so I searched it and it is a subtype of AbstractMatrix so why do we need this new union here?
https://github.com/JuliaLang/julia/blob/670190c3393e7a2193ad9e40cfa12ec643ceba75/stdlib/LinearAlgebra/src/qr.jl#L519

dkarrasch replied:
Because of https://github.com/JuliaLang/julia/pull/46196. This is in preparation of that PR. It makes no difference on older Julia versions, but will on v1.10.

dkarrasch replied:
I should perhaps split the two things in two separate PRs to not mix it up...

JeffFessler replied:
Oh wow, I see.  Thanks for the explanation.  I am not looking forward to having to add more such `Union's` to support things that are kind of like a `AbstractMatrix` (in terms of multiplying) but not a subtype of it.  I've subscribed to the discussion of 46196 so I can see how that debate goes...  Now I'll try to see if this branch works for me.

dkarrasch commented:
Sorry I messed a few things up while splitting. Seems fixed now.

JeffFessler commented:
Ok, I did a test of `LinearMapsAA.jl` and it passed, but with only one deprecation warning so maybe that package does not actually use so many `FunctionMap`s.  So then I also tested `MIRT.jl`and again it passed with exactly one deprecation warning. Maybe that means that `@deprecate` warns only once ( So I guess that bodes well.  I suspect that many of my uses of FMs are spread out in many Literate examples that are not so easy to test but we'll find out in time...

I have not done much (if any) testing of PR branches of others so I want to describe the approach so that you can confirm if I did it properly.  I started Julia 1.8 in an empty directory, then:
```julia
] activate .
add LinearMaps#dk/iipfunctionmap
add LinearMapsAA
test LinearMapsAA
add MIRT
test MIRT
```

So LTGM overall, and I am sure that #193 will bite me too at some point so I am in favor of this direction!

dkarrasch commented:
Thank you very much for the detailed report! If you received deprecation warnings (no matter how many), then it grabbed the right branch. I'm glad that tests pass, too. I hope that `FunctionMap` construction is not done in hot loops, so the deprecation warnings should not yield a performance penalty. I guess this means that it is fine to release this with a minor version bump. For the future, I believe this will allow for some memory optimizations, even though we don't want to dive too deep in there.

fredrikekre commented:
A bit of a drive-by comment, but I personally find `FunctionMap{T,true}(...)` to be less readable compared to `FunctionMap{T}(..., ismutating=true)`. I tried to look at the code, but didn't see where the new parameter is used for dispatch. Does the parameter improve something external or is it just to get rid of `if`s internally?

dkarrasch commented:
Drive-by comments are always welcome! There is dispatch on `IIPFunctionMap` vs. `OOPFunctionMap` in linearcombinations.jl and composed.jl. And yes, it's used to catch cases where all `FunctionMap`s allocate the result array themselves. I don't have a strong preference, it just reminded me that this has been standard in the DiffEq world for a long time, but it could be solved equally well via branching internally I guess. Any thoughts?

JeffFessler commented:
> less readable compared to `FunctionMap{T}(..., ismutating=true)`

I tend to agree.  And the `kwarg` approach has the advantage of being amenable to splatting `kwargs...`
I prefer that approach for the `T=` argument too, but I probably am an outlier there.
I am talking about an outer constructor only; I realize the type itself and the inner constructor could have all sorts of parameters like `FunctionMap{T, F1, F2, iip}(...)`

dkarrasch commented:
Hm, another option could be to make `LinearMaps.ismutating` a "trait" for all `LinearMap` subtypes. That would allow asking a `LinearCombination`/`CompositeMap` whether all its maps are mutating, whereas the current design only catches cases where out-of-place action is due to `FunctionMap`s. But other, custom map types could have the same, and we would miss that. And we could even ask individual maps and do something with that information.

dkarrasch commented:
I realized we already have some kind of the necessary machinery: `MulStyle`, which so far has only `ThreeArg` and `FiveArg`, but which could be extended to `TwoArg`, meaning `A*x`. Reflecting in-place vs. out-of-place behaviour _reasonably_ via `MulStyle`, however, requires that this information is available in the type domain. So how about adding the type parameter `iip`, but not deprecate the old constructors? The only issue with that is that the `iip` parameter may not be inferred from the kwarg constructors. If, however, between map construction and extensive usage is a function barrier, that should be no problem. Thoughts?

JeffFessler commented:
I think it is fine not to have type inference for the kwarg constructors because (at least in my use cases) construction is done just once in an initialization phase and not in any tight loop.  Maintaining the old constructors sounds convenient to me.

dkarrasch commented on test/linearcombination.jl:
```diff
@@ -10,6 +11,18 @@ using LinearMaps: FiveArg, LinearMapTuple, LinearMapVector
     mul!(u, CS!, v)
     @test (@allocated mul!(u, CS!, v)) == 0
     n = 10
+    alloc = @allocated similar(v)
+    Loop = @inferred CS + CS + CS
+    @test Loop * v ≈ 3cumsum(v)
+    @test (@allocated Loop * v) <= 3alloc
+    Loop = @inferred CS + CS; Loop * v
+    @test (@allocated Loop * v) <= 2alloc
+    Lmix = @inferred CS + CS + CS!; Lmix * v
+    @test (@allocated Lmix * v) <= 3alloc
+    Lmix = @inferred CS + CS + CS!; Lmix * v
+    @test (@allocated Lmix * v) <= 3alloc
+    Lmix = @inferred CS! + (CS + CS); Lmix * v
+    @test (@allocated Lmix * v) <= 3alloc
```
These tests are all sharp. They fail on the currently released version!

dkarrasch commented:
This is now new-feature-only, nothing breaks (except for the lack of complete type inference for the convenience constructor). Thanks @fredrikekre for retriggering some thought, thanks @JeffFessler for your comments and testing!

dkarrasch commented:
BTW, I changed the default values of `dest` and `source` in `_compositemul!` to `nothing`, and moved the allocation to the function body (if necessary). We don't use these temporary arrays internally, and any external user will take care of the right size and eltype themselves.
